packages/expo-app-loading/src/AppLoading.tsx

```diff
@@ -0,0 +1,69 @@
+import React from 'react';
+
+import { _emitEvent } from './LifecycleEmitter';
+import * as SplashScreen from './SplashScreen';
+
+export { getAppLoadingLifecycleEmitter } from './LifecycleEmitter';
+
+const AppLoading: React.FC<AppLoadingProps> = props => {
+  const startLoadingResources = React.useCallback(async () => {
+    if (!props.startAsync) {
+      return;
+    }
+
+    if (!props.onFinish) {
+      throw new Error('AppLoading onFinish prop is required if startAsync is provided');
+    }
+
+    try {
+      await props.startAsync();
+    } catch (error) {
+      props.onError!(error);
+    } finally {
+      props.onFinish();
+    }
+  }, []); // eslint-disable-line react-hooks/exhaustive-deps
```

Comment:
We can also adhere to these linter rules. But if the user misuses this, as I encountered with `useFonts`, there is a possibility where multiple `startAsync`s are executed (possibly messing up the preloading).

Reply:
The issue more or less originates from inline functions. If you use `AppLoading` like this, nothing will go wrong:

```
function App() {
  const onAssetsLoad = useCallback(...);
  const onAssetsFinished = useCallback(...);

  return <AppLoading onStart={onAssetsLoad} onFinish={onAssetsFinished} />;
}
```

But since we documented `<AppLoading />` with inline functions, and developers aren't always using callbacks, I like to think most developers actually use it like:

```
function App() {
  return (
    <AppLoading
      onStart={() => ...}
      onFinish={() => ...}
    />
  );
}
```

Which will result in new callback functions every render, possibly messing up the preloading.

Reply:
The caller can partially address this with `useCallback` to memoize the props, but I think the reality is that the hook does not become any less correct by specifying `[]` for its dependencies. That is, the lint rule triggers false positives for legitimate cases and we probably should turn it off. (If a lint rule is helpful but has false positives except in very rare cases, we don't include it in our configuration.)